File: idealingua/idealingua-runtime-rpc-go/src/main/resources/runtime/go/irt/dispatcher.go
package irt

import (
	"fmt"
)

type ServiceDispatcher interface {
	Dispatch(context interface{}, method string, data []byte) ([]byte, error)
	GetSupportedService() string
	GetSupportedMethods() []string
}

type Dispatcher struct {
	services map[string]ServiceDispatcher
}

func (d *Dispatcher) Register(dispatcher ServiceDispatcher) error {
	if dispatcher == nil {
		return fmt.Errorf("trying to register a nil dispatcher")
	}

	if d.services == nil {
		d.services = map[string]ServiceDispatcher{}
	}

	serviceName := dispatcher.GetSupportedService()
	_, exists := d.services[serviceName]
	if exists {
		return fmt.Errorf("trying to register a dispatcher for service %s which is already registered", serviceName)
	}

	d.services[serviceName] = dispatcher
	return nil
}

func (d *Dispatcher) Unregister(serviceName string) bool {
	if d.services == nil {
		return false
	}
	_, ok := d.services[serviceName]
	if ok {
		delete(d.services, serviceName)
		return true
	}

	return false
}

func (d *Dispatcher) Dispatch(context interface{}, service string, method string, data []byte) (out []byte, derr error) {
	if d.services == nil {
		return nil, fmt.Errorf("no services registered to dispatch to")
	}

	dispatcher, ok := d.services[service]
	if !ok {
		return nil, fmt.Errorf("no %s service dispatcher registered", service)
	}

	defer func() {
		if err := recover(); err != nil {
			out = []byte{}
			derr = fmt.Errorf("error in Dispatch for %s/%s: %+v", service, method, err)
		}
	}()

	return dispatcher.Dispatch(context, method, data)
}

File: idealingua/idealingua-runtime-rpc-typescript/src/main/resources/runtime/typescript/irt/dispatcher.ts

export interface ServiceDispatcher<C, D> {
    dispatch(context: C, method: string, data: D | undefined): Promise<D>;
    getSupportedService(): string;
    getSupportedMethods(): string[];
}

export class Dispatcher<C, D> {
    private services: {[key: string]: ServiceDispatcher<C, D>};
    constructor() {
        this.services = {};
    }

    public register(dispatcher: ServiceDispatcher<C, D>): boolean {
        const svc = dispatcher.getSupportedService();
        if (svc in this.services) {
            return false;
        }

        this.services[svc] = dispatcher;
        return true;
    }

    public unregister(serviceName: string): boolean {
        if (!(serviceName in this.services)) {
            return false;
        }

        delete this.services[serviceName];
        return true;
    }

    public dispatch(ctx: C, service: string, method: string, data: D | undefined): Promise<D> {
        if (!(service in this.services)) {
            throw new Error(`Service '${service}' is not registered with the dispatcher.`);
        }

        return this.services[service].dispatch(ctx, method, data);
    }
}
